File: AnkiDroid/src/main/java/com/ichi2/widget/WidgetStatus.java
/***************************************************************************************
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/

package com.ichi2.widget;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.util.Pair;

import com.ichi2.anki.AnkiDroidApp;
import com.ichi2.anki.services.NotificationService;
import com.ichi2.async.BaseAsyncTask;
import com.ichi2.anki.DecksMetaData;

import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import timber.log.Timber;

/**
 * The status of the widget.
 */
public final class WidgetStatus {

    private static boolean sSmallWidgetEnabled = false;
    @SuppressWarnings("deprecation") // #7108: AsyncTask
    private static android.os.AsyncTask<Context, Void, Context> sUpdateDeckStatusAsyncTask;


    /** This class should not be instantiated. */
    private WidgetStatus() {
    }


    /**
     * Request the widget to update its status.
     */
    @SuppressWarnings("deprecation") // #7108: AsyncTask
    public static void update(Context context) {
        SharedPreferences preferences = AnkiDroidApp.getSharedPrefs(context);
        sSmallWidgetEnabled = preferences.getBoolean("widgetSmallEnabled", false);
        boolean canExecuteTask = ((sUpdateDeckStatusAsyncTask == null) || (sUpdateDeckStatusAsyncTask.getStatus() == android.os.AsyncTask.Status.FINISHED));
        if (sSmallWidgetEnabled && canExecuteTask) {
            Timber.d("WidgetStatus.update(): updating");
            sUpdateDeckStatusAsyncTask = new UpdateDeckStatusAsyncTask();
            sUpdateDeckStatusAsyncTask.execute(context);
        } else {
            Timber.d("WidgetStatus.update(): already running or not enabled");
        }
    }


    /** Returns the card due and eta of all the due decks. */
    public static int[] fetchSmall(Context context) {
        DecksMetaData metaData = new DecksMetaData(context);
        Pair<Integer, Integer> pair = metaData.getTotalDueCards();
        return new int[] {pair.first, pair.second};
    }

    private static class UpdateDeckStatusAsyncTask extends BaseAsyncTask<Context, Void, Context> {

        @Override
        protected Context doInBackground(Context... params) {
            super.doInBackground(params);
            Timber.d("WidgetStatus.UpdateDeckStatusAsyncTask.doInBackground()");
            Context context = params[0];
            if (!AnkiDroidApp.isSdCardMounted() && sSmallWidgetEnabled) {
                return context;
            }
            new AnkiDroidWidgetSmall.UpdateService().doUpdate(context);

            // Shows the notification when widget is not set on Home Screen
            Intent intent = new Intent(NotificationService.INTENT_ACTION);
            Context appContext = context.getApplicationContext();
            LocalBroadcastManager.getInstance(appContext).sendBroadcast(intent);
            return context;
        }
    }
}

File: AnkiDroid/src/main/java/com/ichi2/widget/WidgetStatus.kt
// SPDX-License-Identifier: GPL-3.0-or-later

package com.ichi2.widget

import android.content.Context
import com.ichi2.anki.CollectionManager.withCol
import com.ichi2.anki.R
import com.ichi2.anki.common.preferences.sharedPrefs
import com.ichi2.anki.common.utils.android.SdCard
import com.ichi2.anki.common.utils.ext.allDecksCounts
import com.ichi2.anki.settings.Prefs
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import timber.log.Timber

/**
 * The status of the widget.
 */
object WidgetStatus {
    private var smallWidgetEnabled = false
    private var smallWidgetUpdateJob: Job? = null

    /**
     * Request the widget to update its status.
     * TODO Mike - we can reduce battery usage by widget users by removing updatePeriodMillis from metadata
     *             and replacing it with an alarm we set so device doesn't wake to update the widget, see:
     *             https://developer.android.com/guide/topics/appwidgets/#MetaData
     */
    fun updateInBackground(context: Context) {
        val preferences = context.sharedPrefs()
        smallWidgetEnabled = preferences.getBoolean("widgetSmallEnabled", false)
        val canExecuteTask = smallWidgetUpdateJob == null || smallWidgetUpdateJob?.isActive == false

        if (Prefs.newReviewRemindersEnabled) {
            if (smallWidgetEnabled && canExecuteTask) {
                Timber.d("WidgetStatus.update(): updating")
                smallWidgetUpdateJob = launchSmallWidgetUpdateJob(context)
            } else {
                Timber.d("WidgetStatus.update(): already running or not enabled")
            }
        } else {
            val notificationEnabled =
                preferences
                    .getString(context.getString(R.string.pref_notifications_minimum_cards_due_key), "1000001")!!
                    .toInt() < 1000000
            if ((smallWidgetEnabled || notificationEnabled) && canExecuteTask) {
                Timber.d("WidgetStatus.update(): updating")
                smallWidgetUpdateJob = launchSmallWidgetUpdateJob(context)
            } else {
                Timber.d("WidgetStatus.update(): already running or not enabled; enabled: %b", smallWidgetEnabled)
            }
        }
    }

    @OptIn(DelicateCoroutinesApi::class)
    private fun launchSmallWidgetUpdateJob(context: Context): Job =
        GlobalScope.launch {
            try {
                updateSmallWidgetStatus(context)
                Timber.v("launchUpdateJob completed")
            } catch (exc: java.lang.Exception) {
                Timber.w(exc, "failure in widget update")
            }
        }

    suspend fun updateSmallWidgetStatus(context: Context) {
        if (!SdCard.isMounted) {
            Timber.w("updateStatus failed: no SD Card")
            return
        }
        val status = querySmallWidgetStatus()
        widgetRepository.storeSmallWidgetStatus(status)
        if (smallWidgetEnabled) {
            Timber.i("triggering small widget UI update")
            AnkiDroidWidgetSmall.UpdateService().doUpdate(context)
        }
        if (!Prefs.newReviewRemindersEnabled) {
            widgetNotificationScheduler.scheduleNotification()
        }
    }

    /** Returns the status of each of the decks.  */
    fun fetchSmall(): SmallWidgetStatus = widgetRepository.getWidgetSmallStatus()

    fun fetchDue(): Int = widgetRepository.dueCardsCount()

    private suspend fun querySmallWidgetStatus(): SmallWidgetStatus =
        withCol {
            val total = sched.allDecksCounts()
            val eta = sched.eta(total, false)
            SmallWidgetStatus(total.count(), eta)
        }
}
